tonic/src/client.rs

```diff
@@ -0,0 +1,75 @@
+use tonic::service::Interceptor;
+use tonic::transport::Channel;
+
+use celestia_proto::celestia::blob::v1 as blob;
+use celestia_proto::cosmos::auth::v1beta1 as auth;
+use celestia_proto::cosmos::base::node::v1beta1 as config;
+use celestia_proto::cosmos::base::tendermint::v1beta1 as tendermint;
+use celestia_proto::cosmos::tx::v1beta1 as tx;
+use celestia_tendermint_proto::v0_34::types::BlobTx;
+use celestia_types::auth::{AuthParams, BaseAccount};
+use celestia_types::blob::BlobParams;
+
+use crate::types::tx::{GetTxResponse, TxResponse};
+use crate::types::Block;
+use crate::types::{FromGrpcResponse, IntoGrpcParam};
+use crate::Error;
+
+pub struct GrpcClient<I>
+where
+    I: Interceptor,
+{
+    grpc_channel: Channel,
+    auth_interceptor: I,
+}
+
+// macro takes a path to an appropriate generated gRPC method and a desired function signature.
+// If parameters need to be converted, they should implement [`types::IntoGrpcParam`] into
+// appropriate type and return type is converted using TryFrom
+//
+// One limitation is that it expects gRPC method to be provided in exactly 4 `::` delimited
+// segments, requiring importing the proto module with `as`. This might be possible to overcome
+// by rewriting the macro as tt-muncher, but it'd increase its complexity significantly
+macro_rules! make_method2 {
+    ($path:ident :: $client_module:ident :: $client_struct:ident :: $method:ident; $name:ident ( $( $param:ident : $param_type:ty ),* ) -> $ret:ty) => {
+        pub async fn $name(&mut self, $($param: $param_type),*) -> $ret {
+            let mut client = $path::$client_module::$client_struct::with_interceptor(
+                self.grpc_channel.clone(),
+                self.auth_interceptor.clone(),
+            );
+            let request = ::tonic::Request::new(( $($param),* ).into_parameter());
+            let response = client.$method(request).await;
+
+            Ok(response?.into_inner().try_from_response()?)
+        }
+    };
+}
```

Comment:
```suggestion
macro_rules! make_method {
    ($grpc_client:ty, $original_method:ident, $renamed_method:ident( $($param:ident : $param_type:ty),* ) -> $ret:ty) => {
        pub async fn $renamed_method(&mut self, $($param: $param_type),*) -> $ret {
            let mut client = <$grpc_client>::with_interceptor(
                self.grpc_channel.clone(),
                self.auth_interceptor.clone(),
            );
            let request = ::tonic::Request::new(( $($param),* ).into_parameter());
            let response = client.$original_method(request).await;

            response?.into_inner().try_from_response()
        }
    };
}
```

I was thinking about removing this `::` pathing requirement, something like this seems to work. The invocation changes a bit tho, requires a placeholder generic, like:
```rs
    make_method!(config::service_client::ServiceClient<_>, config, get_min_gas_price() -> Result<f64, Error>);
```

Reply:
I've replaced this macro with procedural macro in `grpc_method`